### PR TITLE
Do not crash with internal error when /etc/hosts is corrupted

### PR DESCRIPTION
**Pull request by mchf**
Bnc1077435

extends https://github.com/yast/yast-network/pull/603 some parts should be backported

### CODE REVIEW

File: src/modules/Host.rb [hunk 1]
```diff
@@ -358,12 +358,23 @@ def set_names(address, names)
...
-    @hosts = CFA::Hosts.new
-    @hosts.load
+    begin
+      @hosts = CFA::Hosts.new
+      @hosts.load
+
+      # save hosts to check for changes later
+      @hosts_init = CFA::Hosts.new
+      @hosts_init.load
+    rescue RuntimeError => error
```
Comment (jreidinger) [inline src/modules/Host.rb]: this won't help with IOErrors like file missing. I think it makes sense to add it also there.

File: src/modules/Lan.rb [hunk 2]
```diff
@@ -348,38 +348,48 @@ def Read(cache)
...
+        Builtins.sleep(sl)
+
+        return false if Abort()
+        # Progress step 8/8
+        ProgressNextStage(_("Detecting current status...")) if @gui
+        NetworkService.Read
+        Builtins.sleep(sl)
+
+        return false if Abort()
+      rescue RuntimeError => error
```
Comment (jreidinger) [inline src/modules/Lan.rb]: same IOError can happen here like if you try to write to nfs that is temporary unavailable

File: src/modules/Host.rb (hunk 1 above)
Comment (mchf) [reply src/modules/Host.rb]: This https://github.com/yast/yast-network/pull/621/files#diff-4be40ec54e4c9b91ae249f9931c14eb8R359 should help for missing files

Comment (mchf) [reply src/modules/Host.rb]: However adding IOException is definitely good idea

Comment (jreidinger) [reply src/modules/Host.rb]: it not just about missing files, but also other temporary IO problems that can happen. and it is really `IOError` not IOException

File: src/modules/Lan.rb [hunk 3]
```diff
@@ -348,38 +345,47 @@ def Read(cache)
...
+
+        return false if Abort()
+        # Progress step 8/8
+        ProgressNextStage(_("Detecting current status...")) if @gui
+        NetworkService.Read
+        Builtins.sleep(sl)
+
+        return false if Abort()
+      rescue IOError, SystemCallError, RuntimeError => error
+        msg = format(_("Network configuration is corrupted.\n"\
```
Comment (teclator) [inline src/modules/Lan.rb]: FYI there is already a wrap_text helper https://github.com/search?q=org%3Ayast+wrap_text&type=Code that you could use for wrapping text with a specific width.

Comment (mchf) [reply src/modules/Lan.rb]: thx

File: src/modules/Host.rb [hunk 4]
```diff
@@ -358,12 +358,23 @@ def set_names(address, names)
   def load_hosts
     return false if SCR.Read(path(".target.size"), CFA::Hosts::PATH) <= 0
 
-    @hosts = CFA::Hosts.new
-    @hosts.load
+    begin
```
Comment (mvidner) [inline src/modules/Host.rb]: Did you know you can `rescue` the whole method body?
```rb
def foo
  bar
  baz
rescue E
  qux
end
```

It works just as well here, and the diff+history will be clearer.

File: src/modules/Host.rb [hunk 5]
```diff
@@ -358,12 +358,23 @@ def set_names(address, names)
...
+      @hosts = CFA::Hosts.new
+      @hosts.load
+
+      # save hosts to check for changes later
+      @hosts_init = CFA::Hosts.new
+      @hosts_init.load
+    rescue IOError, SystemCallError, RuntimeError => error
```
Comment (mvidner) [inline src/modules/Host.rb]: I would prefer rescuing their common superclass, which is StandardError.

I know I said "RuntimeError is too generic" in the original PR, but that was in the context of a global Yast handler for such things where I would introduce a GarbageDataError. And I forgot to do it myself when [changing CFA](https://github.com/config-files-api/config_files_api/pull/23/files#diff-53b7f088c9b3245fd0e594323ab213e8R377)

In this context, enumerating the exceptions is confusing to me as a reader, wondering whether we are letting some other exceptions slip through intentionally. Are we? Add a comment if yes, use StandardError if no.

File: src/modules/Lan.rb [hunk 6]
```diff
@@ -348,38 +348,47 @@ def Read(cache)
...
+        Builtins.sleep(sl)
+
+        return false if Abort()
+        # Progress step 8/8
+        ProgressNextStage(_("Detecting current status...")) if @gui
+        NetworkService.Read
+        Builtins.sleep(sl)
+
+        return false if Abort()
+      rescue IOError, SystemCallError, RuntimeError => error
```
Comment (mvidner) [inline src/modules/Lan.rb]: Same as in load_hosts

File: src/modules/Lan.rb [hunk 7]
```diff
@@ -348,38 +348,47 @@ def Read(cache)
       return false if Abort()
       # Progress step 4/8
       ProgressNextStage(_("Reading network configuration...")) if @gui
-      NetworkConfig.Read
+      begin
```
Comment (mvidner) [inline src/modules/Lan.rb]: I wonder why only a part of this long method is worth rescuing. Split off a method with a telling name? Rescue the whole method?

File: src/modules/Lan.rb [hunk 8]
```diff
@@ -348,38 +348,47 @@ def Read(cache)
...
+        NetworkService.Read
+        Builtins.sleep(sl)
+
+        return false if Abort()
+      rescue IOError, SystemCallError, RuntimeError => error
+        msg = format(_("Network configuration is corrupted.\n"\
+                "If you continue resulting configuration can be malformed."\
+                "\n\n%s"), wrap_text(error.message))
+        ret = @gui && Popup.ContinueCancel(msg)
+        return false if !ret
```
Comment (mvidner) [inline src/modules/Lan.rb]: I find this logic hard to understand, I would write
```rb
  return false unless @gui
  return false unless ContinueCancel...
```
but it's your call.

File: src/modules/Lan.rb (hunk 7 above)
Comment (mchf) [reply src/modules/Lan.rb]: ```Lan::Read``` mixes GUI preparation, GUI handling and configuration reading all together. I wanted to catch only exceptions related to configuration reading.

File: src/modules/Lan.rb (hunk 8 above)
Comment (mchf) [reply src/modules/Lan.rb]: ok

File: src/modules/Host.rb (hunk 5 above)
Comment (mchf) [reply src/modules/Host.rb]: added comment

File: src/modules/Host.rb (hunk 4 above)
Comment (mchf) [reply src/modules/Host.rb]: ok

**Pull request merged**